src/Utility/ChatTools.cs

```diff
@@ -0,0 +1,287 @@
+using System;
+using System.ClientModel.Primitives;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json;
+using System.Threading.Tasks;
+using OpenAI.Agents;
+using OpenAI.Chat;
+using OpenAI.Embeddings;
+
+namespace OpenAI.Chat;
+
+/// <summary>
+/// Provides functionality to manage and execute OpenAI function tools for chat completions.
+/// </summary>
+//[Experimental("OPENAIMCP001")]
+public class ChatTools
+{
+    private readonly Dictionary<string, MethodInfo> _methods = [];
+    private readonly Dictionary<string, Func<string, BinaryData, Task<BinaryData>>> _mcpMethods = [];
+    private readonly List<ChatTool> _tools = [];
+    private readonly EmbeddingClient _client;
+    private readonly List<VectorDatabaseEntry> _entries = [];
+    private readonly List<McpClient> _mcpClients = [];
+    private readonly Dictionary<string, McpClient> _mcpClientsByEndpoint = [];
+
+    /// <summary>
+    /// Initializes a new instance of the ChatTools class with an optional embedding client.
+    /// </summary>
+    /// <param name="client">The embedding client used for tool vectorization, or null to disable vectorization.</param>
+    public ChatTools(EmbeddingClient client = null)
+    {
+        _client = client;
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the ChatTools class with the specified tool types.
+    /// </summary>
+    /// <param name="tools">Additional tool types to add.</param>
+    public ChatTools(params Type[] tools) : this((EmbeddingClient)null)
+    {
+        foreach (var t in tools)
+            AddFunctionTool(t);
+    }
+
+    /// <summary>
+    /// Gets the list of defined tools.
+    /// </summary>
+    public IList<ChatTool> Tools => _tools;
+
+    /// <summary>
+    /// Gets whether tools can be filtered using embeddings provided by the provided <see cref="EmbeddingClient"/> .
+    /// </summary>
+    public bool CanFilterTools => _client != null;
+
+    /// <summary>
+    /// Adds local tool implementations from the provided types.
+    /// </summary>
+    /// <param name="tools">Types containing static methods to be used as tools.</param>
+    public void AddFunctionTools(params Type[] tools)
+    {
+        foreach (Type functionHolder in tools)
+            AddFunctionTool(functionHolder);
+    }
+
+    /// <summary>
+    /// Adds all public static methods from the specified type as tools.
+    /// </summary>
+    /// <param name="tool">The type containing tool methods.</param>
+    internal void AddFunctionTool(Type tool)
+    {
+#pragma warning disable IL2070

```

Comment:
We should not be suppressing trimmer warnings like this. The assembly claims to be AOT compatible, but as written this code isn't.

Reply:
We are intentionally defering making this preview feature fully AOT compliant and will address this if it progresses to a GA release.